#405 Allow to apply the AC tool config upon startup

ghenzler opened the pull request:
Closes #405 #407 #408

ghenzler committed:
#405 Allow to apply the AC tool config upon startup

kwin commented on accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/helper/runtime/RuntimeHelper.java:
```diff
@@ -0,0 +1,76 @@
+        try {
+            Node appsNode = session.getNode("/apps");
+            // see https://issues.apache.org/jira/browse/OAK-6563
+            boolean isCompositeNode = !session.hasCapability("addNode", appsNode, new Object[] { "nt:folder" });
```
This could also mean that the given session just does not have the according rights, isn't it? That should be explicitly mentioned or a check for the according rights must be added here as well. 
For simplicity reasons I would always rely on the service user internally and not pass the session as argument....

ghenzler replied:
> For simplicity reasons I would always rely on the service user internally and not pass the session as argument....

The session can be either from the install hook (packmgr) or the service session  - I really want to make sure that the exact session that is also used for installation is used here

> This could also mean that the given session just does not have the according rights, isn't it? 

Correct, will try a combination with `session.hasPermission()` to ensure it's 100% correct

ghenzler replied:
Improved implementation: https://github.com/Netcentric/accesscontroltool/blob/6db01125d0c038cacc2df27b908ad709b129e026/accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/helper/runtime/RuntimeHelper.java#L41

kwin commented on accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/helper/runtime/RuntimeHelper.java:
```diff
@@ -0,0 +1,76 @@
+        return isCloudReady;
+    }
+    
+    public static Node copyNode(Node src, Node dstParent) throws RepositoryException {
```
Can't you use https://docs.adobe.com/docs/en/spec/jsr170/javadocs/jcr-2.0/javax/jcr/Workspace.html#copy(java.lang.String,%20java.lang.String)

kwin commented on accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/helper/runtime/RuntimeHelper.java:
```diff
@@ -0,0 +1,76 @@
+        return dst;
+    }
+
+    public static Property copyProperty(Property src, Node dstParent) throws RepositoryException {
```
IMHO not necessary

kwin commented on accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/impl/AcConfigChangeTracker.java:
```diff
@@ -0,0 +1,81 @@
+        return md5(buf.toString());
+    }
+    
+    public String md5(String input) throws Exception {
```
prrivate?

ghenzler replied:
will fix this

kwin commented on accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/impl/AcInstallationServiceImpl.java:
```diff
@@ -126,16 +130,15 @@
             id = CONFIG_PID)
     protected static @interface Configuration {
-        @AttributeDefinition(name="Configuration storage path", description="CRX path where ACE configuration gets stored")
+        @AttributeDefinition(name="Configuration path", description="JCR path where the AC Tool YAML configuration files are stored")
```
We should mention "root path" and also add some clarification around traversal.

kwin commented on accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/impl/AcInstallationServiceImpl.java:
```diff
@@ -207,22 +222,38 @@ public InstallationLog apply(String configurationRootPath, String[] restrictedTo
     }
 
-    /** Common entry point for JMX and install hook, 
-     * TODO: should not be exported as using non-API class PersistableInstllationLogger */
```
This TODO should be kept!

ghenzler replied:
@kwin please add an issue here (TODOs get forgotten, issues are better in that respect)

kwin replied:
There is already an issue: https://github.com/Netcentric/accesscontroltool/issues/394.
Todo's don't hurt though because they have a better context!

kwin commented on accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/jmx/AceServiceMBean.java:
```diff
@@ -45,8 +45,12 @@ String applyRestrictedToPaths(
             @Name("paths") @Description("comma separated list of paths to apply the configuration to, other paths will be skipped") String restrictedToPaths);
 
-    @Description("Purges the AccessControlList of the given path, if existing")
+    @Description("Applies the ACE configuration if it has not changed")
```
"if it has not  changed"?

kwin commented on accesscontroltool-install-packages/accesscontroltool-cr-package/pom.xml:
```diff
@@ -0,0 +1,98 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
please don't use abbreviations no one will understand. Why not accesscontroltool-cloud-package? The official name is now AEM in the cloud!

ghenzler replied:
`cr` stands for cloud-ready, but yes, `accesscontroltool-cloud-package` is probably better to understand, will adjust it.

ghenzler committed:
#405 startup hook - improvements after review and more testing

ghenzler committed:
#405 re-add TODO

ghenzler committed:
#405 Improved JMX label

ghenzler committed:
#405 Improved install log and better names for Metatype configs

ghenzler committed:
#405 Back to using one package for cloud and local/on-prem

(possible after release of
org.apache.sling.installer.provider.installhook 1.1.0)

ghenzler committed:
#407 Use commons lang 3

ghenzler committed:
#408 Ensure correct ordering of execution history in JMX/WebConsole
